File: lib/backtest/metrics.ts
import { Trade, PerformanceMetrics } from '@/types/backtest';

/**
 * 백테스트 성과 지표 계산
 */
export function calculateMetrics(
  trades: Trade[],
  equity: { time: number; value: number }[],
  initialCapital: number
): PerformanceMetrics {
  if (equity.length === 0) {
    return {
      totalReturn: 0,
      cagr: 0,
      mdd: 0,
      winRate: 0,
      sharpeRatio: 0,
      totalTrades: 0,
    };
  }

  const finalValue = equity[equity.length - 1].value;
  const totalReturn = ((finalValue - initialCapital) / initialCapital) * 100;

  // 거래 쌍 분석 (매수-매도)
  const tradePairs = getTradePairs(trades);
  const winRate = calculateWinRate(tradePairs);

  // MDD 계산
  const mdd = calculateMDD(equity);

  // CAGR 계산
  const cagr = calculateCAGR(equity, initialCapital);

  // 샤프 비율 계산
  const sharpeRatio = calculateSharpeRatio(equity);

  return {
    totalReturn: roundTo(totalReturn, 2),
    cagr: roundTo(cagr, 2),
    mdd: roundTo(mdd, 2),
    winRate: roundTo(winRate, 2),
    sharpeRatio: roundTo(sharpeRatio, 2),
    totalTrades: tradePairs.length,
  };
}

/**
 * 거래 쌍 추출 (매수-매도)
 */
function getTradePairs(trades: Trade[]): { buy: Trade; sell: Trade }[] {
  const pairs: { buy: Trade; sell: Trade }[] = [];

  for (let i = 0; i < trades.length - 1; i += 2) {
    if (trades[i].type === 'buy' && trades[i + 1]?.type === 'sell') {
      pairs.push({ buy: trades[i], sell: trades[i + 1] });
    }
  }

  return pairs;
}

/**
 * 승률 계산
 */
function calculateWinRate(tradePairs: { buy: Trade; sell: Trade }[]): number {
  if (tradePairs.length === 0) return 0;

  const winningTrades = tradePairs.filter(
    (pair) => pair.sell.price > pair.buy.price
  );

  return (winningTrades.length / tradePairs.length) * 100;
}

/**
 * MDD (Maximum Drawdown) 계산
 */
function calculateMDD(equity: { time: number; value: number }[]): number {
  let maxValue = equity[0].value;
  let maxDrawdown = 0;

  for (const point of equity) {
    if (point.value > maxValue) {
      maxValue = point.value;
    }

    const drawdown = ((maxValue - point.value) / maxValue) * 100;
    if (drawdown > maxDrawdown) {
      maxDrawdown = drawdown;
    }
  }

  return maxDrawdown;
}

/**
 * CAGR (연평균 성장률) 계산
 */
function calculateCAGR(
  equity: { time: number; value: number }[],
  initialCapital: number
): number {
  if (equity.length < 2) return 0;

  const startTime = equity[0].time;
  const endTime = equity[equity.length - 1].time;
  const finalValue = equity[equity.length - 1].value;

  // 연 단위로 변환 (타임스탬프는 초 단위)
  const years = (endTime - startTime) / (365.25 * 24 * 60 * 60);

  if (years <= 0) return 0;

  const cagr =
    (Math.pow(finalValue / initialCapital, 1 / years) - 1) * 100;

  return cagr;
}

/**
 * 샤프 비율 계산 (간소화 버전)
 * 일일 수익률의 평균 / 표준편차 * sqrt(252)
 */
function calculateSharpeRatio(equity: { time: number; value: number }[]): number {
  if (equity.length < 2) return 0;

  // 일일 수익률 계산
  const dailyReturns: number[] = [];
  for (let i = 1; i < equity.length; i++) {
    const returnRate =
      (equity[i].value - equity[i - 1].value) / equity[i - 1].value;
    dailyReturns.push(returnRate);
  }

  if (dailyReturns.length === 0) return 0;

  // 평균 수익률
  const meanReturn =
    dailyReturns.reduce((sum, r) => sum + r, 0) / dailyReturns.length;

  // 표준편차
  const variance =
    dailyReturns.reduce((sum, r) => sum + Math.pow(r - meanReturn, 2), 0) /
    dailyReturns.length;
  const stdDev = Math.sqrt(variance);

  if (stdDev === 0) return 0;

  // 샤프 비율 (무위험 수익률은 0으로 가정)
  const sharpeRatio = (meanReturn / stdDev) * Math.sqrt(252);

  return sharpeRatio;
}

/**
 * 소수점 반올림
 */
function roundTo(value: number, decimals: number): number {
  const factor = Math.pow(10, decimals);
  return Math.round(value * factor) / factor;
}

/**
 * 일별 수익률 계산
 */
export function calculateDailyReturns(
  equity: { time: number; value: number }[]
): { time: number; return: number }[] {
  const returns: { time: number; return: number }[] = [];

  for (let i = 1; i < equity.length; i++) {
    const returnRate =
      ((equity[i].value - equity[i - 1].value) / equity[i - 1].value) * 100;

    returns.push({
      time: equity[i].time,
      return: returnRate,
    });
  }

  return returns;
}

File: backend/app/services/metrics.py
"""Performance metrics calculation using NumPy."""

import numpy as np

from app.models.backtest import PerformanceMetrics, Trade


def calculate_metrics(
    trades: list[Trade],
    equity: list[dict[str, float]],
    initial_capital: float,
) -> PerformanceMetrics:
    """
    Calculate backtest performance metrics.

    Args:
        trades: List of trade records
        equity: List of equity points {time, value}
        initial_capital: Starting capital

    Returns:
        PerformanceMetrics object
    """
    if not equity:
        return PerformanceMetrics(
            total_return=0,
            cagr=0,
            mdd=0,
            win_rate=0,
            sharpe_ratio=0,
            total_trades=0,
        )

    equity_values = np.array([e["value"] for e in equity], dtype=np.float64)
    equity_times = np.array([e["time"] for e in equity], dtype=np.int64)

    final_value = equity_values[-1]
    total_return = ((final_value - initial_capital) / initial_capital) * 100

    # Win rate calculation
    win_rate = _calculate_win_rate(trades)

    # MDD calculation
    mdd = _calculate_mdd(equity_values)

    # CAGR calculation
    cagr = _calculate_cagr(equity_times, equity_values, initial_capital)

    # Sharpe ratio
    sharpe = _calculate_sharpe_ratio(equity_values)

    return PerformanceMetrics(
        total_return=round(total_return, 2),
        cagr=round(cagr, 2),
        mdd=round(mdd, 2),
        win_rate=round(win_rate, 2),
        sharpe_ratio=round(sharpe, 2),
        total_trades=len(trades),
    )


def _calculate_win_rate(trades: list[Trade]) -> float:
    """Calculate win rate from trades."""
    # Check for cost_basis based calculation
    sell_trades_with_cost = [t for t in trades if t.type == "sell" and t.cost_basis is not None]

    if sell_trades_with_cost:
        winning = sum(1 for t in sell_trades_with_cost if t.price > (t.cost_basis or 0))
        return (winning / len(sell_trades_with_cost)) * 100

    # Traditional pair-based calculation
    pairs = _get_trade_pairs(trades)
    if not pairs:
        return 0

    winning = sum(1 for buy, sell in pairs if sell.price > buy.price)
    return (winning / len(pairs)) * 100


def _get_trade_pairs(trades: list[Trade]) -> list[tuple[Trade, Trade]]:
    """Extract buy-sell pairs from trades using FIFO matching."""
    pairs: list[tuple[Trade, Trade]] = []
    buy_stack: list[Trade] = []

    for trade in trades:
        if trade.type == "buy":
            buy_stack.append(trade)
        elif trade.type == "sell" and buy_stack:
            pairs.append((buy_stack.pop(0), trade))

    return pairs


def _calculate_mdd(equity_values: np.ndarray) -> float:
    """
    Calculate Maximum Drawdown using NumPy vectorization.

    Args:
        equity_values: Array of equity values

    Returns:
        MDD as percentage
    """
    if len(equity_values) == 0:
        return 0.0

    # Running maximum
    running_max = np.maximum.accumulate(equity_values)

    # Guard against zero running_max (edge case)
    safe_max = np.where(running_max == 0, 1, running_max)

    # Drawdown at each point
    drawdowns = (running_max - equity_values) / safe_max * 100

    return float(np.max(drawdowns))


def _calculate_cagr(
    times: np.ndarray,
    values: np.ndarray,
    initial_capital: float,
) -> float:
    """
    Calculate Compound Annual Growth Rate.

    Args:
        times: Array of timestamps
        values: Array of equity values
        initial_capital: Starting capital

    Returns:
        CAGR as percentage
    """
    if len(times) < 2:
        return 0.0

    start_time = times[0]
    end_time = times[-1]
    final_value = values[-1]

    # Years (timestamps are in seconds)
    years = (end_time - start_time) / (365.25 * 24 * 60 * 60)

    if years <= 0:
        return 0.0

    cagr = (np.power(final_value / initial_capital, 1 / years) - 1) * 100
    return float(cagr)


def _calculate_sharpe_ratio(equity_values: np.ndarray) -> float:
    """
    Calculate Sharpe Ratio (simplified, risk-free rate = 0).

    Args:
        equity_values: Array of equity values

    Returns:
        Annualized Sharpe ratio
    """
    if len(equity_values) < 2:
        return 0.0

    # Daily returns (vectorized)
    daily_returns = np.diff(equity_values) / equity_values[:-1]

    if len(daily_returns) == 0:
        return 0.0

    mean_return = np.mean(daily_returns)
    std_dev = np.std(daily_returns)

    if std_dev == 0:
        return 0.0

    # Annualized Sharpe (assuming daily data, 252 trading days)
    sharpe = (mean_return / std_dev) * np.sqrt(252)
    return float(sharpe)


